File: entity-meta-generator-lib/src/main/java/com/loudsight/meta/serialization/transform/EnumEntityTransform.java
package com.loudsight.meta.serialization.transform;

import com.loudsight.meta.MetaRepository;
import com.loudsight.meta.serialization.EntityTransform;
import com.loudsight.meta.serialization.EntityType;

import java.util.Iterator;
import java.util.List;
import java.util.Map;

public class EnumEntityTransform extends EntityTransform<Object> {

    private static class EnumEntityTransformHolder {
        private static final EnumEntityTransform INSTANCE = new EnumEntityTransform();
    }
    // global access point
    public static EnumEntityTransform getInstance() {
        return EnumEntityTransform.EnumEntityTransformHolder.INSTANCE;
    }
    private EnumEntityTransform() {
        super(EntityType.ENUM, Enum.class);
    }


    @Override public void serializeEntity( Object entity, List<Byte> bytes) {
        bytes.add(EntityType.ENUM.getCode());

        var meta = MetaRepository.getInstance().getMeta(entity.getClass());

        writeStr(meta.getTypeName(), bytes);
        writeStr(((Enum<?>)entity).name(), bytes);
    }

    @Override public Object  deserializeEntity(Iterator<Byte> bytes) {
        var typeName = readStr(bytes);
        var meta = MetaRepository.getInstance().<Enum<?>>getMeta(typeName);
        var enumName = readStr(bytes);

        if (meta == null) {
            // Graceful degrade: return bare enum-name String
            // This occurs on persistence-server for business types not on classpath
            return enumName;
        }

        return meta.newInstance(Map.of("name", enumName));
    }
}

File: entity-meta-generator-lib/src/main/kotlin/com/loudsight/meta/serialization/transform/EnumEntityTransform.kt
package com.loudsight.meta.serialization.transform

import com.loudsight.meta.MetaRepository
import com.loudsight.meta.serialization.EntityTransform
import com.loudsight.meta.serialization.EntityType

object EnumEntityTransform : EntityTransform<Enum<*>>(EntityType.ENUM, Enum::class) {

    override fun serializeEntity(entity: Enum<*>, bytes: MutableList<Byte>) {
        bytes.add(EntityType.ENUM.code)

        val meta = MetaRepository.getMeta(entity::class.java)

        writeStr(meta!!.typeName, bytes)
        writeStr(entity.name, bytes)
    }

    override fun deserializeEntity(bytes: Iterator<Byte>): Enum<*> {
        val typeName = readStr(bytes)
        val meta = MetaRepository.getMeta<Enum<*>>(typeName)
        val enumName = readStr(bytes)

        return meta!!.newInstance(
            mapOf(
                Pair("name", enumName),
            )
        )
    }

    override fun canTransform(entity: Any): Boolean {
        return entity is Enum<*>
    }
}
